design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,486 @@
+- Introduce latencies on node restart.
+- It presents a single point of failure in case the DHCP server is failing
+- It depends on network setup that can be complex.
+
+To overcome the above-mentioned drawbacks, statically configured interfaces taking the address from a
+dynamic host IP pool is a solution. Since the interfaces are static there is no
+reconfiguration after some time or at node restart.
+
+This functionality was requested on U/S via https://github.com/nmstate/kubernetes-nmstate/issues/725
+
+Although this proposal will present a solution with an agnostic storage
+mechanism, following is a spike done using whereabouts at the storage
+https://github.com/nmstate/kubernetes-nmstate/pull/1081
+
+
+### Similar product
+
+- Metal3 IPAM: https://metal3.io/blog/2020/07/06/IP_address_manager.html
+  - It depends on CAPI (ClusterAPI)
+  - It's though for Metal3 baremetal.
+- Azure private addresses: https://docs.microsoft.com/en-us/azure/virtual-network/ip-services/private-ip-addresses
+  - Specific to azure
+  - Some code has to be put in place to consume it.
+
+### User Stories
+
+- As a cluster administrator, I want to configure static IPs on secondary NICs on a pool of nodes without creating one per each node.
+- As a cluster administrator, I want scale up the cluster with new nodes getting IPs from the policy.
+- As a cluster administrator, I want each of these static IPs to be unique, and taken from a defined range, and freed when unused.
+
+### Goals
+
+- Allow users to use NodeNetworkConfigurationPolicy to configure static cluster-unique IP addresses without hard-coding them.
+- If at all possible, we should pick a solution that could be utilized at the baremetal IPI host networkConfig API [1].
+
+[1] For baremetal IPI installation the openshift installer has a api that allow
+to pass Nmstate configuration that will be rendered into NetworkManager keyfiles.
+So it's really the IPI installer config networkData API.
+https://github.com/openshift/installer/blob/969378dc498336d1ac29baa89f75324a44467d5d/pkg/types/baremetal/platform.go#L43
+
+### Non-Goals
+
+- This design does not cover IP address configuration of management NIC on day-0/1
+
+## Proposal
+
+**cluster admin** is a user responsible for managing the cluster node
+networking.
+
+### Workflow Description (assign static address with IPPool)
+
+1. The cluster admin creates `NodeIPPool` resources with IP ranges and a name.
+2. The cluster admin configures node network interfaces statically, with IPs from the `NodeIPPool` using a
+`NodeNetworkConfigurationPolicy` referencing them.
+3. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed
+
+#### Variation (exhausted range at the NodeIPPool)
+
+When the IP pool does not have an available IP address, the `NNCP` will fail
+with a explicit message, the `NodeIPPool` has to increase pool size.
+
+Modifying `NodeIPPool` means increasing CIDR, changing begin/end or exclude,
```

Comment:
What if an already allocated IP is not in the range after the change?

Reply:
+1.

I would also like to see a list of scenarios for which update is possible / not possible.

Keep in mind I think the update scenario is not required, and I have advocated in the past for us to keep the pool's spec immutable.

Reply:
> What if an already allocated IP is not in the range after the change?

The modification is restricted to not exclude already allocated IPs.

> Keep in mind I think the update scenario is not required, and I have advocated in the past for us to keep the pool's spec immutable.

We can stick with NodeIPPool re-create at this design doc and adding update `NodeIPPool` feature later on. @AlonaKaplan @maiqueb what do you think ?

Reply:
+1 to keeping the pool spec immutable.

Reply:
When you re-create the pool. What will happen to the nodes pointing to the old deleted pool?

Reply:
The steps are:
- Create a new NodeIPPool (it has to be in the range of previous one)
- Change NNCPs to reference the new NodeIPPool
- Already reserved IPs will be marked as so in the pool and the node that is missig the IP will get new one.